lib/internal/inspector_async_hook.js

```diff
@@ -16,22 +16,33 @@ const hook = createHook({
     // in https://github.com/nodejs/node/pull/13870#discussion_r124515293,
     // this should be fine as long as we call asyncTaskCanceled() too.
     const recurring = true;
-    inspector.asyncTaskScheduled(type, asyncId, recurring);
+    if (type === 'PROMISE')
+      this.promiseIds.add(asyncId);
+    else
+      inspector.asyncTaskScheduled(type, asyncId, recurring);
   },
 
   before(asyncId) {
+    if (this.promiseIds.has(asyncId))
+      return;
     inspector.asyncTaskStarted(asyncId);
   },
 
   after(asyncId) {
+    if (this.promiseIds.has(asyncId))
+      return;
     inspector.asyncTaskFinished(asyncId);
   },
 
   destroy(asyncId) {
+    if (this.promiseIds.has(asyncId))
+      return this.promiseIds.delete(asyncId);
     inspector.asyncTaskCanceled(asyncId);
   },
 });
 
+hook.promiseIds = new Set();
```

Comment:
Isn't this super implementation dependent? I don't think we are testing or supporting what `this` actually refers to. Is there any advantage of this over `const promiseIds = new Set();`?

Reply:
If we don’t want to guarantee `this` being set to the object passed in (which I think we *should* guarantee, since it’s the most obvious thing), we should set `this` to `undefined` or so. Otherwise other people are going to come to rely on it as well.

Reply:
It was confusing to me because I expected `this` to refer to the options object. That is how `this` works in methods. However, because we reassign the values internally the `this` object becomes something else.

To me, this would be the "obviouse thing": 

```js
 const hook = createHook({
  promiseIds: new Set(),
  init(asyncId, type, triggerAsyncId, resource) {
    console.log(this.promiseIds);
  }
})
```

However, that doesn't work.

Reply:
Having the context be the object passed into `createHooks` would bring symmetry with `Proxy` handlers, but even in the spec community that was a controversial choice (difficulty to extend the range of handlers w/o breaking web compatibility).

Reply:
@TimothyGu Yeah, I feel doing that brings in much more complexity than it removes. `const promiseIds = new Set();` is such a simple solution.